File: Core/src/main/java/com/github/ezauton/core/utils/OddInterpolationMap.java
package com.github.ezauton.core.utils;

import java.util.Map;

/**
 * An interpolation map that has odd symmetry https://en.wikipedia.org/wiki/Even_and_odd_functions
 */
public class OddInterpolationMap extends InterpolationMap {

    public OddInterpolationMap(Double firstKey, Double firstValue) {
        super(firstKey, firstValue);
    }

    public OddInterpolationMap(Map<Double, Double> initTable) {
        super(initTable);
    }

    @Override
    public void putAll(Map<? extends Double, ? extends Double> m) {
        if (m == null) {
            return;
        }
        m.forEach(this::put);
    }

    @Override
    public Double put(Double key, Double value) {
        super.put(-key, -value);
        return super.put(key, value);
    }


}

File: Core/src/main/kotlin/com/github/ezauton/core/utils/OddInterpolationMap.kt
package com.github.ezauton.core.utils

/**
 * An interpolation map that has odd symmetry https://en.wikipedia.org/wiki/Even_and_odd_functions
 */
class OddInterpolationMap : LinearInterpolationMap {

    constructor(firstKey: Double?, firstValue: Double?) : super(firstKey, firstValue)

    constructor(initTable: Map<Double, Double>) : super(initTable)

    override fun putAll(m: Map<out Double, Double>) {
        if (m == null) {
            return
        }
        m.forEach(BiConsumer<out Double, out Double> { key, value -> this.put(key, value) })
    }

    override fun put(key: Double?, value: Double?): Double? {
        super.put((-key)!!, (-value)!!)
        return super.put(key, value)
    }
}
